File: src/main/java/ting/leetcode/java/linkedlist/MiddleOfTheLinkedList.java
package ting.leetcode.java.linkedlist;

public class MiddleOfTheLinkedList {
    public ListNode middleNode(ListNode head) {
        ListNode firstPinter = head;
        ListNode secondPointer = head;

        while (firstPinter != null) {
            firstPinter = firstPinter.next;
            if (firstPinter != null) {
                firstPinter = firstPinter.next;
                secondPointer = secondPointer.next;
            }
        }
        return secondPointer;
    }
}

File: src/main/kotlin/ting/leetcode/beginnerguide/MiddleOfTheLinkedList.kt
package ting.leetcode.beginnerguide

class MiddleOfTheLinkedList {
    fun middleNode(head: ListNode?): ListNode? {
//        var counter = 0
//        var currentNode = head
//        while (currentNode?.next != null) {
//            currentNode = currentNode.next
//            counter++
//        }
//        currentNode = head
//        println(counter)
//        val middleNodeCount = if (counter and 1 == 0) counter / 2 else counter / 2 + 1
//        for (i in 1..(middleNodeCount)) {
//            currentNode = currentNode?.next
//        }
//        return currentNode

//        middleNode will only move forward for 1 node once endNode move 2
        var middleNode = head
        var endNode = head;
        while (endNode?.next != null) {
            middleNode = middleNode?.next
            endNode = endNode.next?.next
        }

        return middleNode
    }
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}